bluefog/common/mpi_controller.cc

```diff
@@ -969,6 +975,113 @@ Status MPIController::WinMutexRelease(const std::string& name,
   return MPIWinMutexReleaseImpl(mutex_win, release_ranks, mpi_ctx_.rank_, is_sync);
 }
 
+Status MPIController::WinVersionGetUpdate(const std::string& name,
+                                          const std::vector<int>& ranks) {
+  BFLOG(TRACE, mpi_ctx_.rank_) << "Update Win Version for " << name << ".";
+
+  auto it = mpi_ctx_.named_win_map.find(name);
+  if (it == mpi_ctx_.named_win_map.end()) {
+    return Status::PreconditionError(
+        "Cannot accquire Version Win for " + name +
+        ". It may not be created or has "
+        "been destroyed or wrong name for associated window.");
+  }
+  std::shared_ptr<MPI_Win> version_win = it->second->GetVersionWin();
+  if (!version_win) {
+    return Status::PreconditionError("Cannot accuire Version Win for " + name +
+                                     ". The data window for that name is found"
+                                     "but the version window is not.");
+  }
+
+  MPI_Win_lock(MPI_LOCK_EXCLUSIVE, mpi_ctx_.rank_, MPI_MODE_NOCHECK,
+                 *version_win);
+  for (int position : ranks) {
+    it->second->updateVersion(position);
+  }
+  MPI_Win_sync(*version_win);
+  MPI_Win_unlock(mpi_ctx_.rank_, *version_win);
+
+  return Status::OK();
+}
+
+Status MPIController::WinVersionPutUpdate(const std::string& name,
+                                          const std::vector<int>& ranks) {
+  BFLOG(TRACE, mpi_ctx_.rank_) << "Update Win Version for " << name << ".";
+
+  auto it = mpi_ctx_.named_win_map.find(name);
+  if (it == mpi_ctx_.named_win_map.end()) {
+    return Status::PreconditionError(
+        "Cannot accquire Version Win for " + name +
+        ". It may not be created or has "
+        "been destroyed or wrong name for associated window.");
+  }
+  std::shared_ptr<MPI_Win> version_win = it->second->GetVersionWin();
+  if (!version_win) {
+    return Status::PreconditionError("Cannot accuire Version Win for " + name +
+                                     ". The data window for that name is found"
+                                     "but the version window is not.");
+  }
+
+  int one = 1;
+
+  for (int rank : ranks) {
+    MPI_Win_lock(MPI_LOCK_SHARED, rank, 0, *version_win);
+    MPI_Accumulate(&one, 1, MPI_INT, rank, /*target_disp=*/mpi_ctx_.rank_, 1, MPI_INT, MPI_SUM,
+                   *version_win);
+    MPI_Win_unlock(rank, *version_win);
+  }
+
+  return Status::OK();
+}
+
+Status MPIController::WinVersionClear(const std::string& name) {
+  BFLOG(TRACE, mpi_ctx_.rank_) << "Win Version for " << name << " is released.";
+  int initial_value = 0;
+
+  auto it = mpi_ctx_.named_win_map.find(name);
```

Comment:
Looks from 
"""
auto it = mpi_ctx_.named_win_map.find(name);
...

std::shared_ptr<MPI_Win> version_win 
""
is repeated four times.[include follwoing get window version value one] (minor different on the PreconditionError).
It should be better to create a private function:
Status GetVersionWinByName(const std::string& name, std::shared_ptr<MPI_Win>& versio_win);

Reply:
As we need to execute the following command:
it->second->resetVersionWinMem();
I think we can leave it in this function.